discovery/azure/azure.go

```diff
@@ -348,10 +360,13 @@ func (d *Discovery) refresh(ctx context.Context) ([]*targetgroup.Group, error) {
 
 			// Get the IP address information via separate call to the network provider.
 			for _, nicID := range vm.NetworkInterfaces {
-				networkInterface, err := client.getNetworkInterfaceByID(ctx, nicID)
-				if err != nil {
-					level.Error(d.logger).Log("msg", "Unable to get network interface", "name", nicID, "err", err)
-					ch <- target{labelSet: nil, err: err}
+				networkInterface, detailedErr := client.getNetworkInterfaceByID(ctx, nicID)
+				if detailedErr != nil {
+					if detailedErr.StatusCode == 404 {
+						level.Warn(d.logger).Log("msg", "Network interface does not exist", "name", nicID, "err", err)
```

Comment:
I think you need `failuresCount.Inc()` here because the other places you have instrumented don't cover this case.

The only log messages that surfaced during the impaired condition described in #10455 were:
```
azure.go:343 "Unable to get network interface"
refresh.go:98 "Unable to refresh target groups"
```

Also, `failuresCount` is being used in a variety of places in this PR, which means it mixes distinct types of errors. Could we apply a label to distinguish these error types? Or specific metric names for each one?

Reply:
Currently, `failuresCount` is only incremented when the refresh returns and error. In this case, the error is ignored and it continues to collect VMs. So that's why it doesn't increment the failure counter, but logs a warning instead. 

I could add a label for different types of failures, but not sure how useful that would be. Users would have to be very familiar with the various labels in order to know how to respond to specific problems. I thought this would mainly be used to alert someone, who will need to go look at the logs and exact error message to determine the cause and fix it. This is also the pattern used in the service discovery components.

Reply:
This makes sense. Sorry it was my mistake--I missed one of your `failuresCount.Inc()` calls when I first reviewed. We do indeed see the following log statement during the real issue with NIC 404s a couple weeks ago. 
```
"unable to complete Azure service discovery"
```
And yes, it's fine to have just one metric name and no labels; after being alerted, logging would be used to diagnose the exact nature of the failure. Thanks!

We'll get this built this week to try it out.